File: riveredge-frontend/src/apps/kuaiplm/services/mold-sample-order.ts
/**
 * 开模合同 / 打样订单 API（R-15 #71）
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/mold-sample-orders';

export type MoldSampleDocKind = 'mold_contract' | 'sample_order';

export type MoldSampleStatus =
  | 'draft'
  | 'pending'
  | 'approved'
  | 'sealed'
  | 'archived'
  | 'rejected';

export interface MoldSampleOrder {
  id: number;
  uuid: string;
  order_code: string;
  project_id: number;
  project_code: string;
  project_name: string;
  doc_kind: MoldSampleDocKind;
  title: string;
  contract_no?: string | null;
  party_name?: string | null;
  file_uuid?: string | null;
  file_name?: string | null;
  status: MoldSampleStatus;
  remarks?: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  sealed_at?: string | null;
  sealed_by_name?: string | null;
  archived_at?: string | null;
  archived_by_name?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface MoldSampleOrderPayload {
  project_id: number;
  doc_kind: MoldSampleDocKind;
  title: string;
  contract_no?: string | null;
  party_name?: string | null;
  file_uuid?: string | null;
  file_name?: string | null;
  remarks?: string | null;
}

function unwrapList(res: unknown): { items: MoldSampleOrder[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as MoldSampleOrder[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const moldSampleOrderApi = {
  list: async (params: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    doc_kind?: string;
    project_id?: number;
  }) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => {
    const res = await api.get(`${BASE}/${id}`);
    return res as MoldSampleOrder;
  },
  create: async (data: MoldSampleOrderPayload) => {
    const res = await api.post(BASE, data);
    return res as MoldSampleOrder;
  },
  update: async (id: number, data: Partial<Omit<MoldSampleOrderPayload, 'project_id'>>) => {
    const res = await api.put(`${BASE}/${id}`, data);
    return res as MoldSampleOrder;
  },
  submit: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/submit`);
    return res as MoldSampleOrder;
  },
  approve: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/approve`);
    return res as MoldSampleOrder;
  },
  reject: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/reject`);
    return res as MoldSampleOrder;
  },
  seal: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/seal`);
    return res as MoldSampleOrder;
  },
  archive: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/archive`);
    return res as MoldSampleOrder;
  },
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
};

File: riveredge-backend/src/apps/kuaiplm/models/mold_sample_order.py
"""开模合同 / 打样订单（R-15 #71）

文员上传 → 审批（前期采购/总经理由审批流配置）→ 打印用印 → 存档。
"""

from tortoise import fields

from core.models.base import BaseModel


class MoldSampleOrder(BaseModel):
    """开模合同或打样订单。"""

    class Meta:
        table = "apps_kuaiplm_mold_sample_orders"
        table_description = "快研发 - 开模合同与打样订单"
        unique_together = [("tenant_id", "order_code")]
        indexes = [
            ("tenant_id", "project_id"),
            ("tenant_id", "status"),
            ("tenant_id", "doc_kind"),
            ("uuid",),
        ]

    id = fields.IntField(pk=True)
    order_code = fields.CharField(max_length=50, description="合同/订单单号")
    project_id = fields.IntField(description="研发项目ID")
    project_code = fields.CharField(max_length=50, description="项目代号快照")
    project_name = fields.CharField(max_length=200, description="项目名称快照")
    doc_kind = fields.CharField(
        max_length=32,
        description="mold_contract=开模合同 / sample_order=打样订单",
    )
    title = fields.CharField(max_length=200, description="标题")
    contract_no = fields.CharField(max_length=80, null=True, description="外部合同号/订单号")
    party_name = fields.CharField(max_length=200, null=True, description="对方单位")
    file_uuid = fields.CharField(max_length=36, null=True, description="合同/订单文件 UUID")
    file_name = fields.CharField(max_length=200, null=True, description="文件名快照")
    status = fields.CharField(
        max_length=20,
        default="draft",
        description="draft/pending/approved/sealed/archived/rejected",
    )
    remarks = fields.TextField(null=True)
    submitted_at = fields.DatetimeField(null=True)
    approved_at = fields.DatetimeField(null=True)
    sealed_at = fields.DatetimeField(null=True)
    sealed_by = fields.IntField(null=True)
    sealed_by_name = fields.CharField(max_length=100, null=True)
    archived_at = fields.DatetimeField(null=True)
    archived_by = fields.IntField(null=True)
    archived_by_name = fields.CharField(max_length=100, null=True)
    deleted_at = fields.DatetimeField(null=True)
